### PR TITLE
A lot of changes (usage of Table.findOne, run-tests,tenant.test, split auth form, etc... Please see comments for more details

### PR DESCRIPTION
**Pull request by AleksandrSokolov**
0. Find and correct all calls for Table.findOne (correct Where). 
1. Split Authentication settings Form (/useradmin/settings) to the 3 forms by semantic
2. a lot of small changes to avoid js warning like wrong list of arguments, unused variables and etc. Please see this.
3. internationalization & typos
4. Improvements for tenant.tests and as result changes in tenant, pack (please check them) 
5. Improvements in run-tests infrastructure (more CLI options for more control of jest)  
6. more js docs / correct js docs 
7. a lof of todo
8. remove unnecessary declarations

### CODE REVIEW

**Comment by AleksandrSokolov** [comment]
HI @glutamate Tom, also I'm not sure about "Rights Settings" name of chapter

**Comment by AleksandrSokolov** [comment]
@glutamate I need idea how to split  Authentications Settings into few blocks in same screen (in one form). 
image

**Comment by glutamate** [comment]
@AleksandrSokolov add a section_header to the field list like this: https://github.com/saltcorn/saltcorn/blob/13380094a0d7badaf7b0120c6adf0050f476663b/packages/saltcorn-data/base-plugin/viewtemplates/feed.js#L77-L80

**Comment by AleksandrSokolov** [comment]
@glutamate Please review. There a lot of changes - clean up unused constants, bug fixing, small optimizations, more js docs, etc. So review can take time.

**Comment by glutamate** [comment]
@AleksandrSokolov thanks, i am working through this. you caught a lot of bugs!

File: packages/server/routes/view.js [hunk 1]
```diff
@@ -104,7 +103,8 @@ router.post(
       if (sf.required && !query[sf.name]) {
         if (!row) {
           if (!table)
-            table = await Table.findOne(view.table_id || view.exttable_name);
+            // todo check after where change
+            table = await Table.findOne({ id : view.table_id } || { name: view.exttable_name });
```
Comment (glutamate) [inline packages/server/routes/view.js:107]: no this doesn't work because `{}` is truthy

File: packages/saltcorn-data/models/field.ts [hunk 2]
```diff
@@ -538,7 +540,7 @@ class Field implements AbstractField {
 
   validate(whole_rec: any): ResultMessage {
     const type = this.is_fkey ? { name: "Key" } : this.type;
-    let readval = null;
+    let readval;
```
Comment (glutamate) [inline packages/saltcorn-data/models/field.ts:543]: why this change? null is not the same as undefined.

Comment (glutamate) [reply packages/saltcorn-data/models/field.ts:543]: I can't explain why it should be null and I don't know if changing this will break it. But I do know that this code is very fiddly due to different types and different representations coming from the browser.

File: packages/saltcorn-data/models/table.ts [hunk 3]
```diff
@@ -152,24 +152,29 @@ class Table implements AbstractTable {
...
    */
-  static findOne(where: Where | Table): Table | null {
+  static findOne(where: Where | Table ): Table | null {
     if (
       where &&
       ((where.constructor && where.constructor.name === "Table") ||
         where.getRows)
     )
       return <Table>where;
+    // todo remove this logic. But needs find all usages! Also this is recursion that usually adds call function overhead
```
Comment (glutamate) [inline packages/saltcorn-data/models/table.ts:166]: I don't think they should be removed because it is used by plug-ins to find their associated table. It's OK to reduce the usage though

File: packages/saltcorn-data/models/table.ts [hunk 4]
```diff
@@ -152,24 +152,29 @@ class Table implements AbstractTable {
...
     if (typeof where === "string") return Table.findOne({ name: where });
     if (typeof where === "number") return Table.findOne({ id: where });
+    // todo i think here bug. External tables will be found only in one case (bottom if)
```
Comment (glutamate) [inline packages/saltcorn-data/models/table.ts:169]: Where do you think this will be buggy - what scenario do you have in mind?

File: packages/saltcorn-data/models/table.ts [hunk 5]
```diff
@@ -152,24 +152,29 @@ class Table implements AbstractTable {
...
     if (where?.name) {
       const { getState } = require("../db/state");
       const extTable = getState().external_tables[where.name];
       if (extTable) return extTable;
     }
     const { getState } = require("../db/state");
+
+    // todo simplify  - I dont understand why this is too complex. And here potential bug because tables can be not cached!!!
```
Comment (glutamate) [inline packages/saltcorn-data/models/table.ts:177]: It's complex because it's a micro optimisation. It is calculating the anonymous function to pass to find, so we don't have to go through all of these conditionals for every element in the table array.

File: packages/saltcorn-data/models/table.ts [hunk 6]
```diff
@@ -399,7 +405,7 @@ class Table implements AbstractTable {
       for (const trigger of triggers) {
         for (const row of rows) {
           // run triggers on delete
-          await trigger.run!(row);
+          trigger.run!(row);
```
Comment (glutamate) [inline packages/saltcorn-data/models/table.ts:408]: I think we need to wait here?

File: packages/saltcorn-data/models/table.ts [hunk 7]
```diff
@@ -625,7 +635,7 @@ class Table implements AbstractTable {
       const row = await this.getRow({ id });
       if (!row) throw new Error(`Unable to find row with id: ${id}`);
       for (const trigger of triggers) {
-        await trigger.run!(row);
+        trigger.run!(row);
```
Comment (glutamate) [inline packages/saltcorn-data/models/table.ts:638]: again i think we need to await?

**Review by glutamate — CHANGES_REQUESTED** [review]

**Review by glutamate — APPROVED** [review]
